apollo-router/src/plugins/authentication/subgraph_auth.rs

```diff
@@ -0,0 +1,455 @@
+use std::collections::HashMap;
+use std::ops::ControlFlow;
+use std::sync::Arc;
+use std::time::SystemTime;
+
+use aws_credential_types::provider::ProvideCredentials;
+use aws_credential_types::Credentials;
+use aws_sigv4::http_request;
+use aws_sigv4::http_request::sign;
+use aws_sigv4::http_request::PayloadChecksumKind;
+use aws_sigv4::http_request::SignableBody;
+use aws_sigv4::http_request::SignableRequest;
+use aws_sigv4::http_request::SigningSettings;
+use aws_types::region::Region;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use tower::BoxError;
+use tower::ServiceBuilder;
+use tower::ServiceExt;
+
+use crate::layers::ServiceBuilderExt;
+use crate::plugin::Plugin;
+use crate::plugin::PluginInit;
+use crate::register_plugin;
+use crate::services::subgraph;
+use crate::services::SubgraphRequest;
+
+register_plugin!("apollo", "subgraph_authentication", SubgraphAuth);
+
+/// Hardcoded Config using assess_key and secret.
+/// Prefer using DefaultChain instead.
+#[derive(Clone, JsonSchema, Deserialize, Debug)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+struct AWSSigV4HardcodedConfig {
+    /// The ID for this access key.
+    access_key_id: String,
+    /// The secret key used to sign requests.
+    secret_access_key: String,
+    /// The aws region this chain applies to.
+    region: String,
+    /// The service you're trying to access, eg: "s3", "vpc-lattice-svcs", etc.
+    service_name: String,
+    /// Specify a role ARN for your credentials.
+    assume_role: Option<AssumeRoleProvider>,
+}
+
+impl ProvideCredentials for AWSSigV4HardcodedConfig {
+    fn provide_credentials<'a>(
+        &'a self,
+    ) -> aws_credential_types::provider::future::ProvideCredentials<'a>
+    where
+        Self: 'a,
+    {
+        aws_credential_types::provider::future::ProvideCredentials::ready(Ok(Credentials::new(
+            self.access_key_id.clone(),
+            self.secret_access_key.clone(),
+            None,
+            None,
+            "apollo-router",
+        )))
+    }
+}
+
+/// Configuration of the DefaultChainProvider
+#[derive(Clone, JsonSchema, Deserialize, Debug)]
+struct DefaultChainConfig {
+    /// The aws region this chain applies to.
+    region: String,
+    /// The profile name used by this provider
+    profile_name: Option<String>,
+    /// The service you're trying to access, eg: "s3", "vpc-lattice-svcs", etc.
+    service_name: String,
+    /// Specify a role ARN for your credentials.
```

Comment:
`Specify a role ARN` is probably better suited for `AssumeRoleProvider`.
```suggestion
    /// Specify assumed role configuration
```

Reply:
Fixed in 7509a060